File: 3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dictionary = {}
        curr_length = 0
        max_length = 0
        for i in range(len(s)):
            if s[i] in dictionary:
                curr_length = max(curr_length, dictionary[s[i]])
            max_length = max(max_length, i - curr_length + 1)
            dictionary[s[i]] = i + 1
        return max_length


# @lc code=end

File: 3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int n = s.size();
        int res = 0;
        unordered_map<char, int> m;
        for (int i = 0, j = 0; i < n; i++)
        {
            if (m.find(s[i]) != m.end())
            {
                j = max(j, m[s[i]] + 1);
            }
            res = max(res, i - j + 1);
            m[s[i]] = i;
        }
        return res;
    }
};
// @lc code=end
